Libs: mem_blocks library extensions

marcinszkudlinski opened the pull request:
This is an extension of memory blocks library. 
Three functions were added:

- sys_mem_blocks_alloc_continuous
- sys_mem_blocks_free_continuous
- sys_mem_blocks_get

above changes require also changes in bitarray lib:

-  sys_bitarray_test_and_set_region added

marcinszkudlinski committed:
lib/mem_blocks: change internal alloc_one to alloc_blocks

As preparation for memory blocks to support continuous memory blocks
modify bitmap operation wrappers to support multiple bits alloc/free

Signed-off-by: Marcin Szkudlinski <marcin.szkudlinski@intel.com>

marcinszkudlinski committed:
UnitTest: lib: mem_block: bugfix -  table size

listener* tables are in fact some kind of log of all alloc/free
operations. Therefore size of the tables == NUM_BLOCKS makes
no sense and leads to table oveflows.

Signed-off-by: Marcin Szkudlinski <marcin.szkudlinski@intel.com>

marcinszkudlinski committed:
UnitTest: lib: test_mem_block_get test added

Unit tests for new memory_block library sys_mem_blocks_get functionality

Signed-off-by: Marcin Szkudlinski <marcin.szkudlinski@intel.com>

marcinszkudlinski committed:
UnitTest: lib: test_mem_block_get increase num of blocks

As preparation for continous testing there 's a need to increase
number of blocks in tests

Signed-off-by: Marcin Szkudlinski <marcin.szkudlinski@intel.com>

marcinszkudlinski committed:
UnitTests: lib mem_blocks: block_alloc_free_contiguous/free

Unit tests for mem_blocks library block_alloc_free_contiguous/free

Signed-off-by: Marcin Szkudlinski <marcin.szkudlinski@intel.com>

marcinszkudlinski committed:
lib: bitarray: sys_bitarray_test_and_set_region added

a method to check and set/clear a chosen region in a bitmap
if not previously set/cleared in a single atomic operation.
Useful for keeping track of resources usage, like memory banks

Signed-off-by: Marcin Szkudlinski <marcin.szkudlinski@intel.com>

marcinszkudlinski committed:
lib/os: sys_mem_blocks_get added

memory block should allow not only allocate memory but also
getting an arbitrary chosen part of memory.
Mixing of allocate and get, however possible, may be dangerous
because alloc may get any of memory. So be careful.

Signed-off-by: Marcin Szkudlinski <marcin.szkudlinski@intel.com>

marcinszkudlinski committed:
lib/mem_blocks: add allocation for contiguous memory blocks

Added allocation and freeing of a contiguous memory blocks

Signed-off-by: Marcin Szkudlinski <marcin.szkudlinski@intel.com>

andyross commented on lib/os/mem_blocks.c:
```diff
@@ -109,6 +109,52 @@ int sys_mem_blocks_alloc(sys_mem_blocks_t *mem_block, size_t count,
+	int ret = 0;
+	void *ptr;
+
+	if ((mem_block == NULL) || (out_block == NULL)) {
```
Same comment about __ASSERT vs. runtime error for "should never happen" conditions like API misuse by the caller.

andyross commented on lib/os/mem_blocks.c:
```diff
@@ -57,6 +57,45 @@ static int free_bits(sys_mem_blocks_t *mem_block, void *ptr, size_t bits)
+{
+	int ret = 0;
+
+	if ((mem_block == NULL) || (out_block == NULL)) {
```
Same deal: would prefer defense against "obvious bugs in the calling code" to be assertions and not runtime code bloat.

andyross commented on lib/os/mem_blocks.c:
```diff
@@ -57,6 +57,45 @@ static int free_bits(sys_mem_blocks_t *mem_block, void *ptr, size_t bits)
+		goto out;
+	}
+
+	void *ptr = alloc_bits(mem_block, count);
```
I'm too lazy to check: is the existing alloc_bits() utility atomic?  If not, then you should probably add a lock here for symmetry with the other APIs being added.

marcinszkudlinski replied:
it is atomic, in fact it is using an atiomic procedure sys_bitarray_alloc

andyross commented on tests/lib/mem_blocks/src/main.c:
```diff
@@ -205,6 +205,182 @@ static void test_mem_block_multi_alloc_free_alt_buf(void)
+					    i, &val);
+		switch (i) {
+		case 0:
+		case 1:
```
Style: seems like a needlessly complicated switch vs. something like:

```c
if (i < 2) {
	zassert_equal(val, 1,
		      "sys_mem_blocks_get bitmap failed, bit %i should be set", i);
} else {
	zassert_equal(val, 0,
		      "sys_mem_blocks_get bitmap failed, bit %i should be cleared", i);
}
```

marcinszkudlinski replied:
yes and no. I intentionally made it with case, to explicitly show which blocks should be taken/free. It's getting more complicated in other tests (next commit)  - and ifs like in your comment would become hard to read. All of tests and verifications are prepared in the way it would run properly regardless of NUM_BLOCKS value

andyross commented on tests/lib/mem_blocks/src/main.c:
```diff
@@ -205,6 +205,182 @@ static void test_mem_block_multi_alloc_free_alt_buf(void)
+	/* blocks 0 and 1 should be taken */
+	for (i = 0; i < NUM_BLOCKS; i++) {
+		ret = sys_bitarray_test_bit(mem_block_01.bitmap,
+					    i, &val);
```
ret gets assigned but never checked; that's going to show up later as more work for some poor soul doing code coverage analysis.  Just a `zassert_equal(ret, 0, "API failure");` is enough.  This pattern repeats elsewhere too.

marcinszkudlinski replied:
fixed in all tests (not only added ones)

andyross reviewed (commented):
A few nitpicks, looks good in general.

marcinszkudlinski commented on include/sys/mem_blocks.h:
```diff
@@ -218,6 +218,27 @@ struct sys_multi_mem_blocks {
+ *
+ * Allocate a specified blocks in a memory block object.
+ * Note: use caution when mixing sys_mem_blocks_get and sys_mem_blocks_alloc,
+ * allocation may take any of the free memory space
```
Danger is here: once a standard alloc is executed, it may alloc a block any address in the mem_block. Even worse, the address may vary from run to run. So when there's a need to get a memory at a specific address it may be available or not in more/less random conditions. So best practice is: don't mix get and allocs OR do all gets before first alloc.

dcpleung replied:
If that's the case, it's better to put it in the code documentation, possibly in both functions as a note or warning.

marcinszkudlinski commented on lib/os/bitarray.c:
```diff
@@ -547,6 +547,39 @@ static int set_clear_region(sys_bitarray_t *bitarray, size_t num_bits,
+	if ((num_bits == 0)
+	    || (num_bits > bitarray->num_bits)
+	    || (offset >= bitarray->num_bits)
+	    || (off_end >= bitarray->num_bits)) {
```
This is a copy&paste from other bitarray.c procedures, should I change in the whole file?

marcinszkudlinski replied:
@andyross one more detail - ztest cannot handle kernel panic, so I had to remove test_mem_block_invalid_params and test_multi_mem_block_invalid_params tests.

nashif replied:
it sure does, there are many examples of this in tests, for example:

```
START - test_z_phys_map_side_effect
E: Page fault at address 0x7f3000 (error code 0x3)
E: Access violation: supervisor thread not allowed to write
E: PTE: 0x7f3000 -> 0x000000000010d000: XD
E: EAX: 0x007f3000, EBX: 0x00116468, ECX: 0x0010c7b0, EDX: 0x00000001
E: ESI: 0x00000000, EDI: 0x0010137a, EBP: 0x001103c8, ESP: 0x001103c4
E: EFLAGS: 0x00000202 CS: 0x0008 CR3: 0x0011c000
E: call trace:
E: EIP: 0x001003af
E:      0x00103509 (0x116454)
E:      0x00101389 (0x116468)
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x10b080 (test_z_phys_map_side_effect)
Caught system error -- reason 0
 PASS - test_z_phys_map_side_effect in 0.3 seconds
```

This is usually dealt with a fatal handler, for example:

```
void k_sys_fatal_error_handler(unsigned int reason, const z_arch_esf_t *pEsf)
{
        printk("Caught system error -- reason %d\n", reason);

        if (expect_fault && reason == 0) {
                expect_fault = false;
                ztest_test_pass();
        } else {
                printk("Unexpected fault during test");
                k_fatal_halt(reason);
        }
}
```

marcinszkudlinski commented on tests/lib/mem_blocks/src/main.c:
```diff
@@ -381,6 +381,188 @@ static void test_mem_block_get(void)
+		case 4:
+		case 5:
+		case 6:
+		case 7:
```
@andyross I prefer to leave the checks as they are now, it is much clearer what the test is really checking

marcinszkudlinski committed:
os/lib: change all NULL checking in bitarray.c to asserts

Some problems, as passing NULL when a context in expected,
always reflect to bugs. Such problems cannot be handled
in runtime. So the desired action is system panic.

Signed-off-by: Marcin Szkudlinski <marcin.szkudlinski@intel.com>

marcinszkudlinski committed:
os/lib: change all NULL checking in mem_blocks.c to asserts

Some problems, as passing NULL when a context in expected,
always reflect to bugs. Such problems cannot be handled
in runtime. So the desired action is system panic.

Signed-off-by: Marcin Szkudlinski <marcin.szkudlinski@intel.com>

dcpleung commented:
If the tests are expected to generate kernel panics, it is possible to override `k_sys_fatal_error_handler()` to catch the errors and to fail the tests if no exceptions are raised. See `tests/kernel/fatal/exception/src/main.c` for example.

marcinszkudlinski committed:
UnitTest: lib: change unit tests to handle asserts

Change negative tests to handle exceptions
instead of exit codes

Signed-off-by: Marcin Szkudlinski <marcin.szkudlinski@intel.com>

marcinszkudlinski commented:
@dcpleung pls approve again - some checks (docs and commit signoff) failed 
@andyross pls review

marcinszkudlinski commented:
@andyross ?

andyross reviewed (approved):
Refresh review.  Remaining nitpicks shouldn't block merge.
